delivery_carrier_label_ups/models/stock_picking.py

```diff
@@ -398,9 +412,16 @@ def _ups_request(
             "Password": account.password,
             "transactionSrc": "Odoo (%s)" % self.env.cr.dbname,
         }
-        return getattr(requests, method)(
+        res = getattr(requests, method)(
             url, json=payload, headers=headers, params=query_parameters
         ).json()
+        # log request and response
+        ups_last_request = ("URL: {}").format(url)
+        if payload:
+            ups_last_request += ("\nData: {}").format(str(payload))
+        self.carrier_id.log_xml(ups_last_request, "ups_last_request")
+        self.carrier_id.log_xml(str(res.get("response", {})) or "", "ups_last_response")
```

Comment:
This creates empty logs in our setup. It seems that successful requests don't have a `response` entry, but a `ShippingResponse`, or a `LabelRecoveryResponse` etc. I think you can log `res` itself.

Reply:
Thanks, works now!